Library/Notifier.swift

```diff
@@ -4,35 +4,69 @@ import UIKit
-        let dateWithNext30 = Date().andNext30Days()
-        for date in dateWithNext30 {
-            self.scheduleNotification(for: location, at: date)
-        }
-    }
-
-    private static func scheduleNotification(for location: Location, at date: Date) {
-        guard let location = Location.current else { return }
-
+    private static func scheduleNotification(for sunTime: SunTime, at date: Date, withMessage message: String) {
         let notificationID = String(DateHasher.hashValue(for: date))
-        let sunriseDate = location.sunTime.sunriseStartTime(for: date)
+        let sunriseDate = sunTime.sunriseStartTime(for: date)
 
-        let formattedMessage = self.formattedMessage(location: location, date: date)
         let components = sunriseDate.components([.calendar, .year, .month, .day, .hour, .minute, .second, .timeZone])
 
         let trigger = UNCalendarNotificationTrigger(dateMatching: components, repeats: false)
         let content = UNMutableNotificationContent()
-        content.body = formattedMessage
+        content.body = message
         let request = UNNotificationRequest(identifier: notificationID, content: content, trigger: trigger)
 
         UNUserNotificationCenter.current().add(request)
     }
 
-    static func formattedMessage(location: Location, date: Date) -> String {
-        return Message.notificationMessage(for: date, coordinates: location.coordinates)
+    func cancelAllNotifications() {
+        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
     }
 
-    static func cancelAllNotifications() {
-        UNUserNotificationCenter.current().removeAllPendingNotificationRequests()
+    var areNotificationsEnabled: Bool {
+        set {
+            UserDefaults.standard.set(newValue, forKey: "isNotificationsEnabled")
+        }
+
+        get {
+            let userRequestedNotification = UserDefaults.standard.bool(forKey: "isNotificationsEnabled")
+            if userRequestedNotification == false {
+                return false
+            } else {
+                return self.isAllowedToSendNotifications
+            }
+        }
+    }
+
+    func registerForNotifications() {
+        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge]) { _, error in
+            print(error ?? "Registered for notifications.")
+        }
+    }
+
+    var isAllowedToSendNotifications: Bool {
+        return self.notificationAuthorizationStatus() == .authorized
+    }
+
+    func notificationAuthorizationStatus() -> UNAuthorizationStatus {
+        var status = UNAuthorizationStatus.notDetermined
+
+        // Make call synchronous.
+        let semaphore = DispatchSemaphore(value: 0)
+
+        UNUserNotificationCenter.current().getNotificationSettings { settings in
+            status = settings.authorizationStatus
+            semaphore.signal()
+        }
+        _ = semaphore.wait(timeout: DispatchTime.distantFuture)
```

Comment:
I can see this came from someplace else, but I would **very** strongly recommend against this - there's a reason this call isn't synchronous from the system. You can just pass this method a completion block.

Reply:
Although in theory we should have the UI element just show a loading state until we get a response from this method, in my experience just flattening it out like this works pretty well for this particular case. Despite having no assurances that this call will return quickly I've never faced a situation where it didn't take more than a couple CPU cycles to return.

Reply:
There's absolutely no guarantee that will always be the case though. 

My opinion is that if Apple wanted devs to treat this call as synchronous, they'd have made it synchronous.